File: code/2_parameters/09_mortality_in_care.R
suppressMessages(library(geepack))
suppressMessages(library(haven))
suppressMessages(library(R.utils))
suppressMessages(library(tidyverse))
suppressMessages(library(broom))

pearl_dir <- Sys.getenv("PEARL_DIR")
input_dir <- filePath(pearl_dir, '/param/raw')
intermediate_dir <- filePath(pearl_dir, '/param/intermediate')
param_dir <- filePath(pearl_dir, '/param/param')

group_names = c('msm_white_male', 'msm_black_male', 'msm_hisp_male', 'idu_white_male', 'idu_white_female',
                'idu_black_male', 'idu_black_female', 'idu_hisp_male', 'idu_hisp_female', 'het_white_male',
                'het_white_female', 'het_black_male', 'het_black_female', 'het_hisp_male', 'het_hisp_female')

#####################################################################################
# Model 1: Mortality among those in care
#####################################################################################
# Read in analysis population

pop1 <- read_sas(filePath(input_dir, 'pop_mortality.sas7bdat'))
colnames(pop1) <- tolower(colnames(pop1))
pop1$pop2 <- tolower(pop1$pop2)

yearly <- pop1 %>%
  mutate(ageby10 = floor(age/10),
         ageby10 = replace(ageby10, ageby10==1, 2),
         ageby10 = replace(ageby10, ageby10 > 7, 7)) %>%
  filter(year >= 2009, # modified 9/26/18
         year <= 2017) %>% # added 09/11/18
  select(pop2, sex, naid, realdeath, year, ageby10, sqrtcd4n, h1yy) %>%  
  na.omit() %>%
  group_by(pop2, sex) %>% 
  nest()


modelfx <- function(df) {
  mylogit <- geeglm(realdeath ~ year + ageby10 + sqrtcd4n + h1yy, 
                    id = naid, 
                    data = df, 
                    corstr = "exchangeable",
                    family=binomial(link='logit'))
}



get_coeff <- function(DF) {
  coeffs <- coef(DF)
  coeffs <- data.frame(t(coeffs))
  colnames(coeffs) <- c("intercept", "year", "age_cat", "sqrtcd4n", "h1yy")
  return(coeffs)
}

model2 <- yearly %>%
  mutate(model = map(data, modelfx),
         coeffs = map(model, get_coeff),
         vcov = map(model, ~.$geese$vbeta)) %>%
  select(-data)

model2$sex[model2$sex==1] <- "male"
model2$sex[model2$sex==2] <- "female"

model2 <- model2 %>% unite(group, pop2, sex, remove=TRUE)

coeffs <- model2 %>% 
  select(group, coeffs) %>% 
  unnest(cols=coeffs)

vcov <- model2 %>% 
  select(group, vcov) %>% 
  mutate_if(is.list, map, as_tibble, .name_repair='unique') %>%
  unnest(cols=vcov)

write_csv(coeffs, filePath(param_dir, 'mortality_in_care.csv'))
write_csv(vcov, filePath(param_dir, 'mortality_in_care_vcov.csv'))

File: code/2_parameters/09_mortality_in_care.py
# Imports
import os
import pandas as pd
from pathlib import Path

# Define directories
pearl_dir = Path(os.getenv('PEARL_DIR'))
in_dir = f'{pearl_dir}/param/raw/'
out_dir = f'{pearl_dir}/param/param/'

group_names = ['msm_white_male', 'msm_black_male', 'msm_hisp_male', 'idu_white_male', 'idu_black_male',
               'idu_hisp_male', 'idu_white_female', 'idu_black_female', 'idu_hisp_female', 'het_white_male',
               'het_black_male', 'het_hisp_male', 'het_white_female', 'het_black_female', 'het_hisp_female']
group_names.sort()

df = pd.read_csv(f'{in_dir}/mortality_in_care_coeffs.csv')
df.columns = df.columns.str.lower()
df['pop3'] = df['pop3'].str.lower()
df['parm'] = df['parm'].str.lower()
df = df[['pop3', 'parm', 'estimate']]
df1 = df.loc[df['pop3'] == 'idu_hisp + idu_white_women'].copy()
df = df.loc[df['pop3'] != 'idu_hisp + idu_white_women'].copy()
df1 = pd.concat([df1.assign(pop3='idu_hisp_women'), df1.assign(pop3='idu_white_women')])
df = pd.concat([df, df1], ignore_index=True)
df['sex'] = '-'
df.loc[df['pop3'].str.contains('_men'), 'sex'] = 'male'
df.loc[df['pop3'].str.contains('_women'), 'sex'] = 'female'

group_split = df['pop3'].str.split('_', expand=True)
df['group'] = group_split[0] + '_' + group_split[1] + '_' + df['sex']
df = df[['group', 'parm', 'estimate']]
df = df.set_index(['group', 'parm']).sort_index().reset_index()
df = df.pivot_table(index='group', columns='parm', values='estimate').reset_index()
df.to_csv(f'{out_dir}/mortality_in_care.csv', index=False)

df = pd.read_csv(f'{in_dir}/mortality_in_care_knots.csv')
df.columns = df.columns.str.lower()
df['pop2'] = df['pop2'].str.lower()
df['variable'] = df['variable'].str.lower()
df1 = df.loc[df['pop2'] == 'idu_hisp + idu_white'].copy()
df = df.loc[df['pop2'] != 'idu_hisp + idu_white'].copy()
df1 = pd.concat([df1.assign(pop2='idu_hisp'), df1.assign(pop2='idu_white')])
df = pd.concat([df, df1], ignore_index=True)
df.loc[df['sex'] == 1, 'sex'] = 'male'
df.loc[df['sex'] == 2, 'sex'] = 'female'
df['group'] = df['pop2'] + '_' + df['sex']
df = df[['variable', 'group', 'p5', 'p35', 'p65', 'p95']].set_index(['variable', 'group']).sort_index()
df1 = df.loc['age'].rename(columns={'p5': 1, 'p35': 2, 'p65': 3, 'p95': 4}).reset_index()
df1.to_csv(f'{out_dir}/mortality_in_care_age.csv', index=False)
df2 = df.loc['sqrt(cd4 at art ini)'].rename(columns={'p5': 1, 'p35': 2, 'p65': 3, 'p95': 4}).reset_index()
df2.to_csv(f'{out_dir}/mortality_in_care_sqrtcd4.csv', index=False)
print(df2)



